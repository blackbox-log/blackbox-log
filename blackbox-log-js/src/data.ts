import { Memoize as memoize } from 'typescript-memoize';

import { WasmPointer } from './wasm';

import type { Headers } from './headers';
import type { WasmExports, WasmObject } from './wasm';

export class DataParser implements WasmObject {
	readonly #wasm: WasmExports;
	readonly #ptr: WasmPointer;
	readonly #headers: Headers;

	constructor(wasm: WasmExports, headers: Headers, ptr: number) {
		this.#wasm = wasm;
		this.#headers = headers;
		this.#ptr = new WasmPointer(ptr, wasm.data_free);
	}

	free() {
		this.#ptr.free();
	}

	get isAlive(): boolean {
		return this.#ptr.isAlive;
	}

	get headers(): Headers {
		return this.#headers;
	}

	parseAll() {
		this.#wasm.data_parseAll(this.#ptr.ptr);
	}

	get mainFrameCount(): number {
		return this.#wasm.data_mainFrameCount(this.#ptr.ptr);
	}

	get gpsFrameCount(): number {
		return this.#wasm.data_gpsFrameCount(this.#ptr.ptr);
	}
}
