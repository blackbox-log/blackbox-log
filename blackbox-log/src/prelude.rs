//! Minimal set of imports for convenience.
//!
//! [`File`](`crate::File`) is not included due to avoid collisions with other
//! types.

pub use crate::frame::{Frame as _, FrameDef as _};
pub use crate::{DataParser, FieldFilter, Headers, ParserEvent};
