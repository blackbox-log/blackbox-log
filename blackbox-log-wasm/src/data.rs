use blackbox_log::prelude::*;
use blackbox_log::Reader;

use crate::{OwnedSlice, Shared, WasmFfi};

// SAFETY: field order *must* be `parser` first, then `headers`, then `data` to
// ensure correct drop order
pub struct WasmDataParser {
    parser: DataParser<'static, 'static>,
    _headers: Shared<Headers<'static>>,
    _data: Shared<OwnedSlice>,
}

impl WasmDataParser {
    pub(crate) fn new(
        headers: Shared<Headers<'static>>,
        reader: Reader<'static>,
        data: Shared<OwnedSlice>,
    ) -> Self {
        // SAFETY: this is only used to create the `DataParser`, which is guaranteed to
        // be dropped before `headers` by the declaration order in the struct
        let headers_ref = unsafe { headers.deref_static() };

        Self {
            parser: DataParser::new(reader, headers_ref),
            _headers: headers,
            _data: data,
        }
    }

    fn main_frame_count(&self) -> usize {
        self.parser.stats().counts.main
    }

    fn gps_frame_count(&self) -> usize {
        self.parser.stats().counts.gps
    }
}

impl WasmFfi for WasmDataParser {}

#[no_mangle]
unsafe extern "wasm" fn data_free(ptr: *mut WasmDataParser) {
    let parser = WasmDataParser::from_wasm(ptr);
    drop(parser);
}

#[no_mangle]
#[allow(non_snake_case)]
unsafe extern "wasm" fn data_mainFrameCount(ptr: *mut WasmDataParser) -> usize {
    let parser = WasmDataParser::from_wasm(ptr);
    let count = parser.main_frame_count();
    parser.into_wasm();
    count
}

#[no_mangle]
#[allow(non_snake_case)]
unsafe extern "wasm" fn data_gpsFrameCount(ptr: *mut WasmDataParser) -> usize {
    let parser = WasmDataParser::from_wasm(ptr);
    let count = parser.gps_frame_count();
    parser.into_wasm();
    count
}
