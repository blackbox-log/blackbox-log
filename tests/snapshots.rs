use std::collections::BTreeMap;
use std::fs;
use std::io::Read;

use blackbox_log::data::Stats;
use blackbox_log::frame::{self, FieldDef, GpsFrameDef};
use blackbox_log::prelude::*;
use blackbox_log::units::si;
use blackbox_log::{headers, Event, Unit, Value};
use hashbrown::HashMap;
use serde::ser::{SerializeStruct, Serializer};
use serde::Serialize;

const MAX_EVENTS: usize = 200_000;
const MAX_LOGS: usize = 2;

macro_rules! run {
    () => {
        |path| {
            let mut file = fs::File::open(path).unwrap();
            let mut data = Vec::new();
            file.read_to_end(&mut data).unwrap();

            let file = blackbox_log::File::new(&data);
            let snapshot = FileSnapshot::from(file);

            insta::assert_ron_snapshot!(snapshot);
        }
    };
}

#[test]
fn own() {
    insta::glob!("logs/*.bbl", run!());
}

#[test]
#[ignore]
fn fc_blackbox() {
    insta::glob!("logs/fc-blackbox/*", run!());
}

#[test]
#[ignore]
fn gimbal_ghost() {
    insta::glob!("logs/gimbal-ghost/*", run!());
}

#[derive(Debug, Serialize)]
#[serde(rename = "File")]
struct FileSnapshot<'data> {
    count: usize,
    logs: Vec<headers::ParseResult<LogSnapshot<'data>>>,
}

impl<'data> From<blackbox_log::File<'data>> for FileSnapshot<'data> {
    fn from(file: blackbox_log::File<'data>) -> Self {
        let logs = file
            .iter()
            .map(|mut reader| {
                Headers::parse(&mut reader).map(|headers| {
                    let data = DataParser::new(reader, &headers);
                    LogSnapshot::new(&headers, data)
                })
            })
            .take(MAX_LOGS)
            .collect::<Vec<_>>();

        Self {
            count: file.log_count(),
            logs,
        }
    }
}

#[derive(Debug, Serialize)]
#[serde(rename = "Log")]
struct LogSnapshot<'data> {
    headers: HeadersSnapshot<'data>,
    stats: Stats,
    capped: bool,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    events: Vec<Event>,
    #[serde(skip_serializing_if = "MainSnapshot::is_empty")]
    main: MainSnapshot,
    #[serde(skip_serializing_if = "SlowSnapshot::is_empty")]
    slow: SlowSnapshot,
    #[serde(skip_serializing_if = "GpsSnapshot::is_empty")]
    gps: GpsSnapshot,
}

impl<'data> LogSnapshot<'data> {
    fn new(headers: &Headers<'data>, mut data: DataParser<'data, '_>) -> Self {
        let mut events = Vec::new();

        let main = headers.main_frame_def.iter().collect::<Fields>();
        let mut main = MainSnapshot::new(main);

        let slow = headers.slow_frame_def.iter().collect::<Fields>();
        let mut slow = SlowSnapshot::new(slow);

        let gps = headers
            .gps_frame_def
            .iter()
            .flat_map(GpsFrameDef::iter)
            .collect::<Fields>();
        let mut gps = GpsSnapshot::new(gps);

        let mut capped = true;
        for _ in 0..MAX_EVENTS {
            let Some(frame) = data.next() else {
                capped = false;
                break;
            };

            match frame {
                ParserEvent::Event(event) => events.push(event),
                ParserEvent::Main(frame) => main.update(frame),
                ParserEvent::Slow(frame) => slow.update(frame),
                ParserEvent::Gps(frame) => gps.update(frame),
            }
        }

        Self {
            headers: headers.into(),
            stats: data.stats().clone(),
            capped,
            events,
            main,
            slow,
            gps,
        }
    }
}

#[derive(Debug, Serialize)]
#[serde(rename = "Headers")]
struct HeadersSnapshot<'data> {
    firmware_revision: &'data str,
    firmware: headers::Firmware,
    firmware_date: Option<Result<String, &'data str>>,
    board_info: Option<&'data str>,
    craft_name: Option<&'data str>,
    debug_mode: headers::DebugMode,
    disabled_fields: headers::DisabledFields,
    features: headers::FeatureSet,
    unknown: HashMap<&'data str, &'data str>,
}

impl<'data, 'a> From<&'a Headers<'data>> for HeadersSnapshot<'data> {
    fn from(headers: &'a Headers<'data>) -> Self {
        HeadersSnapshot {
            firmware_revision: headers.firmware_revision(),
            firmware: headers.firmware(),
            firmware_date: headers
                .firmware_date()
                .map(|res| res.map(|date| date.to_string())),
            board_info: headers.board_info(),
            craft_name: headers.craft_name(),
            debug_mode: headers.debug_mode(),
            disabled_fields: headers.disabled_fields(),
            features: headers.features(),
            unknown: headers.unknown().clone(),
        }
    }
}

#[derive(Debug, Serialize)]
#[serde(rename = "MainFrame")]
struct MainSnapshot {
    count: u128,
    last_iteration: u32,
    time: NumberHistory<16, i128>,
    fields: Fields,
}

impl MainSnapshot {
    fn new(fields: Fields) -> Self {
        Self {
            count: 0,
            last_iteration: 0,
            time: NumberHistory::new(),
            fields,
        }
    }

    fn update(&mut self, frame: frame::MainFrame) {
        self.count += 1;
        self.last_iteration = frame.iteration();
        self.time.update(frame.time_raw().into());
        self.fields.update(frame);
    }

    fn is_empty(&self) -> bool {
        self.count == 0
    }
}

#[derive(Debug, Serialize)]
#[serde(rename = "SlowFrame")]
struct SlowSnapshot {
    count: u32,
    fields: Fields,
}

impl SlowSnapshot {
    fn new(fields: Fields) -> Self {
        Self { count: 0, fields }
    }

    fn update(&mut self, frame: frame::SlowFrame) {
        self.count += 1;
        self.fields.update(frame);
    }

    fn is_empty(&self) -> bool {
        self.count == 0
    }
}

#[derive(Debug, Serialize)]
#[serde(rename = "GpsFrame")]
struct GpsSnapshot {
    count: u32,
    time: (),
    fields: Fields,
}

impl GpsSnapshot {
    fn new(fields: Fields) -> Self {
        Self {
            count: 0,
            time: (),
            fields,
        }
    }

    fn update(&mut self, frame: frame::GpsFrame) {
        self.count += 1;
        self.fields.update(frame);
    }

    fn is_empty(&self) -> bool {
        self.count == 0
    }
}

#[derive(Debug, Serialize)]
struct Fields(Vec<FieldSnapshot>);

impl Fields {
    fn update<F: blackbox_log::frame::Frame>(&mut self, frame: F) {
        for (field, value) in self.0.iter_mut().zip(frame.iter()) {
            field.update(value.into());
        }
    }
}

impl<'data, U: Into<Unit>> FromIterator<FieldDef<'data, U>> for Fields {
    fn from_iter<I: IntoIterator<Item = FieldDef<'data, U>>>(iter: I) -> Self {
        Self(
            iter.into_iter()
                .map(|field| FieldSnapshot::new(field.name.to_owned(), field.unit.into()))
                .collect(),
        )
    }
}

#[derive(Debug, Serialize)]
#[serde(rename = "Field")]
struct FieldSnapshot {
    name: String,
    unit: Unit,
    history: History,
}

#[derive(Debug, Serialize)]
#[serde(untagged)]
enum History {
    Int(NumberHistory<16, i128>),
    Bool { yes: usize, no: usize },
    Flags(BTreeMap<&'static str, usize>),
}

#[derive(Debug)]
struct NumberHistory<const N: usize, T> {
    min: T,
    max: T,
    seen: Vec<T>,
    histogram: [usize; N],
}

impl FieldSnapshot {
    fn new(name: String, unit: Unit) -> Self {
        Self {
            name,
            unit,
            history: match unit {
                Unit::Amperage
                | Unit::Voltage
                | Unit::Acceleration
                | Unit::Rotation
                | Unit::GpsCoordinate
                | Unit::Altitude
                | Unit::Velocity
                | Unit::GpsHeading
                | Unit::Unitless => History::Int(NumberHistory::new()),
                Unit::Boolean => History::Bool { yes: 0, no: 0 },
                Unit::FlightMode | Unit::State | Unit::FailsafePhase => {
                    History::Flags(BTreeMap::new())
                }
            },
        }
    }

    #[allow(clippy::wildcard_enum_match_arm, clippy::cast_possible_truncation)]
    fn update(&mut self, value: Value) {
        match &mut self.history {
            History::Int(history) => history.update(match value {
                Value::Amperage(a) => a.get::<si::electric_current::milliampere>().round() as i128,
                Value::Voltage(v) => v.get::<si::electric_potential::millivolt>().round() as i128,
                Value::Acceleration(a) => a
                    .get::<si::acceleration::centimeter_per_second_squared>()
                    .round() as i128,
                Value::Rotation(r) => {
                    r.get::<si::angular_velocity::degree_per_second>().round() as i128
                }
                Value::GpsCoordinate(c) => (c * 10000000.).round() as i128,
                Value::Altitude(a) => a.get::<si::length::meter>().round() as i128,
                Value::Velocity(v) => {
                    v.get::<si::velocity::centimeter_per_second>().round() as i128
                }
                Value::GpsHeading(h) => (h * 10.).round() as i128,
                Value::Unsigned(u) => u.into(),
                Value::Signed(s) => s.into(),
                _ => unreachable!(),
            }),
            History::Bool { yes, no } => match value {
                Value::Boolean(true) => *yes += 1,
                Value::Boolean(false) => *no += 1,
                _ => unreachable!(),
            },
            History::Flags(history) => {
                if let Value::FailsafePhase(phase) = value {
                    *history.entry(phase.as_name()).or_insert(0) += 1;
                } else {
                    let flags = match value {
                        Value::FlightMode(m) => m.as_names(),
                        Value::State(s) => s.as_names(),
                        _ => unreachable!(),
                    };

                    for flag in flags {
                        *history.entry(flag).or_insert(0) += 1;
                    }
                }
            }
        }
    }
}

impl<const N: usize, T> NumberHistory<N, T> {
    fn new() -> Self
    where
        T: Default,
    {
        Self {
            min: T::default(),
            max: T::default(),
            seen: Vec::new(),
            histogram: [0; N],
        }
    }

    fn update_range(&mut self, value: T)
    where
        T: PartialOrd,
    {
        if value < self.min {
            self.min = value;
        } else if value > self.max {
            self.max = value;
        }
    }

    /// Store new value sorted in `seen` if not already present
    fn update_seen(&mut self, value: T)
    where
        T: PartialOrd,
    {
        let index = self.seen.partition_point(|x| *x < value);
        if self.seen.get(index) != Some(&value) {
            self.seen.insert(index, value);
        }
    }

    /// Bucket so that sequential values go in sequential buckets, since data is
    /// usually clustered
    #[inline(always)]
    fn update_histogram(&mut self, value: u128) {
        self.histogram[(value % N as u128) as usize] += 1;
    }
}

impl<const N: usize> NumberHistory<N, i128> {
    fn update(&mut self, value: i128) {
        self.update_range(value);
        self.update_seen(value);
        self.update_histogram(value.unsigned_abs());
    }
}

impl<const N: usize> Serialize for NumberHistory<N, i128>
where
    [usize; N]: Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serialize(self, self.min, self.max, serializer)
    }
}

impl<const N: usize> Serialize for NumberHistory<N, f64>
where
    [usize; N]: Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serialize(
            self,
            format!("{:.2}", self.min),
            format!("{:.2}", self.max),
            serializer,
        )
    }
}

#[inline(always)]
fn serialize<const N: usize, T, S, U>(
    snapshot: &NumberHistory<N, T>,
    min: U,
    max: U,
    serializer: S,
) -> Result<S::Ok, S::Error>
where
    S: Serializer,
    U: Serialize,
    [usize; N]: Serialize,
{
    let unique = snapshot.seen.len();

    let mut state = serializer.serialize_struct("FieldSnapshot", 4)?;

    state.serialize_field("min", &min)?;
    state.serialize_field("max", &max)?;
    state.serialize_field("unique", &unique)?;

    let histogram = "histogram";
    if unique > 1 {
        state.serialize_field(histogram, &snapshot.histogram)?;
    } else {
        state.skip_field(histogram)?;
    }

    state.end()
}
