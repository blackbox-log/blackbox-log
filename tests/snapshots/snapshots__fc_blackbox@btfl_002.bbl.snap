---
source: tests/snapshots.rs
expression: snapshot
input_file: ".modules/fc-blackbox/src/test-data/btfl_002.bbl"
---
File(
  count: 1,
  logs: [
    Ok(Log(
      headers: Headers(
        firmware_revision: "Betaflight 4.2.8 (101738d8e) STM32F7X2",
        firmware: Betaflight("4.2.8"),
        firmware_date: Some(Ok("2021-02-15 12:10:35.0")),
        board_info: Some("AIKO AIKONF7"),
        craft_name: Some("Gecko"),
        debug_mode: None,
        disabled_fields: [],
        features: [
          RxSerial,
          ServoTilt,
          Telemetry,
          Osd,
          AirMode,
          EscSensor,
          AntiGravity,
          DynamicFilter,
        ],
        pwm_protocol: Dshot600,
        unknown: {
          "I interval": "256",
          "Log start datetime": "0000-01-01T00:00:00.000+00:00",
          "P interval": "16",
          "P ratio": "16",
          "abs_control_gain": "0",
          "acc_hardware": "0",
          "acc_limit": "0",
          "acc_limit_yaw": "0",
          "acc_lpf_hz": "1000",
          "airmode_activate_throttle": "25",
          "anti_gravity_gain": "3500",
          "anti_gravity_mode": "0",
          "anti_gravity_threshold": "250",
          "baro_hardware": "1",
          "d_min": "27,30,0",
          "d_min_advance": "20",
          "d_min_gain": "37",
          "deadband": "1",
          "dshot_bidir": "1",
          "dshot_idle_value": "550",
          "dterm_filter2_type": "0",
          "dterm_filter_type": "1",
          "dterm_lowpass2_hz": "0",
          "dterm_lowpass_dyn_hz": "0,350",
          "dterm_lowpass_hz": "100",
          "dterm_notch_cutoff": "0",
          "dterm_notch_hz": "0",
          "dterm_rpm_notch_harmonics": "0",
          "dterm_rpm_notch_min": "100",
          "dterm_rpm_notch_q": "500",
          "dyn_notch_max_hz": "500",
          "dyn_notch_min_hz": "100",
          "dyn_notch_q": "350",
          "dyn_notch_width_percent": "5",
          "feedforward_transition": "0",
          "feedforward_weight": "140,170,90",
          "ff_boost": "15",
          "ff_interpolate_sp": "2",
          "ff_max_rate_limit": "100",
          "ff_spike_limit": "60",
          "gyro_cal_on_first_arm": "0",
          "gyro_hardware_lpf": "0",
          "gyro_lowpass2_hz": "0",
          "gyro_lowpass2_type": "0",
          "gyro_lowpass_dyn_hz": "0,400",
          "gyro_lowpass_hz": "300",
          "gyro_lowpass_type": "0",
          "gyro_notch_cutoff": "105,152",
          "gyro_notch_hz": "115,162",
          "gyro_rpm_notch_harmonics": "3",
          "gyro_rpm_notch_min": "120",
          "gyro_rpm_notch_q": "500",
          "gyro_sync_denom": "1",
          "iterm_relax": "1",
          "iterm_relax_cutoff": "15",
          "iterm_relax_type": "1",
          "iterm_windup": "100",
          "levelPID": "50,50,75",
          "looptime": "125",
          "magPID": "40",
          "mag_hardware": "1",
          "maxthrottle": "2000",
          "motor_pwm_rate": "480",
          "pidAtMinThrottle": "1",
          "pid_process_denom": "1",
          "pidsum_limit": "500",
          "pidsum_limit_yaw": "400",
          "pitchPID": "39,80,40",
          "rate_limits": "1998,1998,1998",
          "rates": "70,70,70",
          "rates_type": "3",
          "rc_expo": "60,60,60",
          "rc_interpolation": "2",
          "rc_interpolation_channels": "2",
          "rc_interpolation_interval": "19",
          "rc_rates": "10,10,10",
          "rc_smoothing_active_cutoffs": "100, 80",
          "rc_smoothing_auto_factor": "20",
          "rc_smoothing_cutoffs": "0, 0",
          "rc_smoothing_debug_axis": "0",
          "rc_smoothing_filter_type": "1, 1",
          "rc_smoothing_rx_average": "3998",
          "rc_smoothing_type": "1",
          "rollPID": "35,75,35",
          "rpm_notch_lpf": "150",
          "serialrx_provider": "9",
          "thr_expo": "0",
          "thr_mid": "50",
          "tpa_breakpoint": "1450",
          "tpa_rate": "70",
          "use_integrated_yaw": "0",
          "use_unsynced_pwm": "0",
          "vbat_pid_gain": "0",
          "vbatcellvoltage": "330,350,430",
          "yawPID": "40,60,0",
          "yaw_deadband": "1",
          "yaw_lowpass_hz": "0",
        },
      ),
      stats: Stats(
        counts: FrameCounts(
          event: 4,
          main: 66640,
          slow: 17,
          gps: 0,
          gps_home: 0,
        ),
      ),
      capped: false,
      events: [
        Resume(
          log_iteration: 4608,
          time: 17433272,
        ),
        SyncBeep(16734098),
        FlightMode(
          flags: 524289,
          last_flags: 268435459,
        ),
        Disarm(4),
      ],
      main: MainFrame(
        count: 66640,
        last_iteration: 1070607,
        time: FieldSnapshot(
          min: 0,
          max: 151271525,
          unique: 66640,
          histogram: (4255, 4323, 4397, 4162, 4113, 4234, 2417, 5808, 4173, 3926, 4194, 3954, 4149, 4259, 2258, 6018),
        ),
        fields: Fields([
          Field(
            name: "axisP[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -160,
              max: 118,
              unique: 196,
              histogram: (22746, 16615, 9823, 5578, 3299, 2228, 1634, 1177, 907, 664, 556, 385, 340, 266, 223, 199),
            ),
          ),
          Field(
            name: "axisP[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -271,
              max: 93,
              unique: 145,
              histogram: (23067, 16843, 10114, 5615, 3156, 2002, 1408, 1011, 739, 666, 515, 430, 368, 288, 235, 183),
            ),
          ),
          Field(
            name: "axisP[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -193,
              max: 98,
              unique: 78,
              histogram: (28471, 18571, 8697, 4250, 2307, 1438, 881, 606, 390, 269, 214, 184, 135, 100, 74, 53),
            ),
          ),
          Field(
            name: "axisI[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -43,
              max: 36,
              unique: 80,
              histogram: (9551, 9168, 7405, 6700, 5245, 4623, 3867, 3514, 2903, 3070, 3238, 2231, 1546, 1085, 1410, 1084),
            ),
          ),
          Field(
            name: "axisI[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -92,
              max: 89,
              unique: 168,
              histogram: (4009, 3347, 2793, 2741, 2522, 2661, 2564, 3472, 4607, 5952, 5380, 6265, 5705, 5757, 5310, 3555),
            ),
          ),
          Field(
            name: "axisI[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -69,
              max: 45,
              unique: 115,
              histogram: (6832, 6505, 6352, 6877, 6176, 5266, 4892, 4619, 4316, 3584, 2884, 2435, 1834, 1550, 1238, 1280),
            ),
          ),
          Field(
            name: "axisD[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -240,
              max: 283,
              unique: 318,
              histogram: (9814, 9294, 8156, 7467, 6191, 5119, 4197, 3555, 2803, 2296, 1893, 1601, 1266, 1183, 961, 844),
            ),
          ),
          Field(
            name: "axisD[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -1052,
              max: 367,
              unique: 177,
              histogram: (11929, 10707, 9377, 7963, 6254, 5098, 3859, 2977, 2243, 1715, 1280, 995, 794, 606, 463, 380),
            ),
          ),
          Field(
            name: "axisF[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -208,
              max: 254,
              unique: 313,
              histogram: (36839, 11409, 6214, 3770, 2424, 1544, 1046, 821, 596, 447, 409, 319, 232, 227, 178, 165),
            ),
          ),
          Field(
            name: "axisF[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -45,
              max: 46,
              unique: 80,
              histogram: (41855, 10495, 5084, 2934, 1871, 1322, 868, 622, 441, 325, 239, 180, 138, 125, 73, 68),
            ),
          ),
          Field(
            name: "axisF[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -15,
              max: 15,
              unique: 31,
              histogram: (46146, 10153, 4796, 2375, 1311, 718, 438, 231, 186, 120, 76, 40, 26, 15, 6, 3),
            ),
          ),
          Field(
            name: "rcCommand[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -344,
              max: 499,
              unique: 842,
              histogram: (5827, 3269, 6922, 6372, 5148, 4259, 4109, 3263, 3525, 3784, 2916, 3767, 3611, 2656, 3391, 3821),
            ),
          ),
          Field(
            name: "rcCommand[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -349,
              max: 211,
              unique: 557,
              histogram: (6893, 2452, 5030, 7122, 6658, 4090, 4020, 2977, 3548, 4402, 2741, 3340, 3316, 2845, 3287, 3919),
            ),
          ),
          Field(
            name: "rcCommand[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -325,
              max: 328,
              unique: 654,
              histogram: (3385, 2945, 4292, 4249, 7229, 5004, 6569, 4404, 3458, 3557, 3431, 3577, 4046, 2928, 3802, 3764),
            ),
          ),
          Field(
            name: "rcCommand[3]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 2000,
              unique: 808,
              histogram: (3920, 4303, 4562, 2520, 3697, 3965, 3778, 3650, 8867, 4055, 3772, 3735, 4276, 4191, 4153, 3196),
            ),
          ),
          Field(
            name: "setpoint[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -221,
              max: 702,
              unique: 785,
              histogram: (8067, 9607, 4304, 4976, 3988, 3927, 3868, 4154, 3266, 3542, 2550, 2797, 3124, 2807, 2791, 2872),
            ),
          ),
          Field(
            name: "setpoint[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -229,
              max: 87,
              unique: 305,
              histogram: (7533, 13396, 4487, 5485, 3268, 4901, 4215, 4266, 2826, 3205, 2217, 2504, 2476, 2112, 1749, 2000),
            ),
          ),
          Field(
            name: "setpoint[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -194,
              max: 198,
              unique: 391,
              histogram: (4482, 13455, 4258, 4751, 3667, 3978, 3684, 3400, 3287, 3830, 3020, 3013, 2847, 3671, 3261, 2036),
            ),
          ),
          Field(
            name: "setpoint[3]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 1000,
              unique: 771,
              histogram: (9027, 3884, 3838, 3680, 3870, 3521, 3991, 3774, 3925, 3574, 3502, 3899, 3654, 3682, 4172, 4647),
            ),
          ),
          Field(
            name: "vbatLatest",
            unit: Voltage,
            history: FieldSnapshot(
              min: 0,
              max: 16580,
              unique: 308,
              histogram: (8539, 0, 8613, 0, 8089, 0, 8216, 0, 8445, 0, 7793, 0, 8234, 0, 8711, 0),
            ),
          ),
          Field(
            name: "amperageLatest",
            unit: Amperage,
            history: FieldSnapshot(
              min: 0,
              max: 0,
              unique: 1,
            ),
          ),
          Field(
            name: "rssi",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 680,
              unique: 238,
              histogram: (3435, 4720, 4480, 4584, 4286, 4034, 4034, 3740, 5582, 3934, 4085, 4133, 3638, 3638, 4036, 4281),
            ),
          ),
          Field(
            name: "gyroADC[0]",
            unit: Rotation,
            history: FieldSnapshot(
              min: -234,
              max: 730,
              unique: 851,
              histogram: (5701, 7850, 6257, 5341, 4946, 4399, 4210, 3935, 3635, 3364, 3105, 2878, 2718, 2861, 2810, 2630),
            ),
          ),
          Field(
            name: "gyroADC[1]",
            unit: Rotation,
            history: FieldSnapshot(
              min: -234,
              max: 201,
              unique: 338,
              histogram: (7187, 9768, 7495, 5484, 4539, 4310, 4105, 3747, 3323, 2860, 2577, 2467, 2347, 2225, 2094, 2112),
            ),
          ),
          Field(
            name: "gyroADC[2]",
            unit: Rotation,
            history: FieldSnapshot(
              min: -196,
              max: 200,
              unique: 397,
              histogram: (6181, 8244, 6164, 5073, 4504, 4182, 3830, 3573, 3506, 3291, 3161, 3042, 3035, 3110, 3029, 2715),
            ),
          ),
          Field(
            name: "accSmooth[0]",
            unit: Acceleration,
            history: FieldSnapshot(
              min: -4231,
              max: 204,
              unique: 965,
              histogram: (4188, 4181, 4229, 4134, 4108, 4169, 4122, 4183, 4269, 4201, 4049, 4143, 4371, 4267, 4087, 3939),
            ),
          ),
          Field(
            name: "accSmooth[1]",
            unit: Acceleration,
            history: FieldSnapshot(
              min: -748,
              max: 1090,
              unique: 1176,
              histogram: (4429, 4485, 4641, 4440, 4408, 4245, 4163, 4092, 3968, 3980, 3951, 4185, 3939, 3928, 3909, 3877),
            ),
          ),
          Field(
            name: "accSmooth[2]",
            unit: Acceleration,
            history: FieldSnapshot(
              min: -266,
              max: 7527,
              unique: 4017,
              histogram: (4095, 4224, 4116, 4159, 4182, 4201, 4234, 3991, 4159, 4052, 4217, 4339, 4157, 4124, 4227, 4163),
            ),
          ),
          Field(
            name: "motor[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 2055,
              unique: 1417,
              histogram: (4197, 4096, 4089, 4158, 4054, 4150, 4141, 4067, 4121, 4218, 4001, 4147, 4075, 4371, 4645, 4110),
            ),
          ),
          Field(
            name: "motor[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 2046,
              unique: 1487,
              histogram: (4195, 4105, 4025, 4194, 4152, 4154, 4186, 4164, 4158, 4185, 4088, 4210, 4170, 4270, 4217, 4167),
            ),
          ),
          Field(
            name: "motor[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 2075,
              unique: 1408,
              histogram: (4141, 4119, 4022, 4079, 4068, 4028, 4024, 4036, 4155, 4092, 4142, 4218, 4513, 4723, 4109, 4171),
            ),
          ),
          Field(
            name: "motor[3]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 2038,
              unique: 1488,
              histogram: (4079, 4159, 4094, 4090, 4106, 4133, 4079, 4177, 4127, 4224, 4108, 4243, 4149, 4461, 4239, 4172),
            ),
          ),
        ]),
      ),
      slow: SlowFrame(
        count: 17,
        fields: Fields([
          Field(
            name: "flightModeFlags",
            unit: FlightMode,
            history: {
              "ARM": 17,
              "BLACKBOX": 17,
            },
          ),
          Field(
            name: "stateFlags",
            unit: State,
            history: {},
          ),
          Field(
            name: "failsafePhase",
            unit: FailsafePhase,
            history: {
              "IDLE": 17,
            },
          ),
          Field(
            name: "rxSignalReceived",
            unit: Boolean,
            history: History(
              yes: 17,
              no: 0,
            ),
          ),
          Field(
            name: "rxFlightChannelsValid",
            unit: Boolean,
            history: History(
              yes: 17,
              no: 0,
            ),
          ),
        ]),
      ),
    )),
  ],
)
